### PR TITLE
feat: add docker support

### PR DESCRIPTION
**Pull request by bestony**
# What does it do?
add Docker Support for Hexo Cli.

## How to test
1. install Docker
2.
```sh
git clone -b add-dockerfile https://github.com/bestony/hexo.git
cd hexo
./hexod # use this to show hexo info 
./hexod init testdocker
```

## Other info 
1. About hexo in Docker Hub

is [hexo](https://hub.docker.com/u/hexo) belongs to hexo?

Now, i push the image to my docker hub account, If this belongs to Hexo Official, I can push image to Hexo Organization.

## Screenshots

image

image

## Pull request tasks

- [ ] Add test cases for the changes.
- [ ] Passed the CI test.

### CODE REVIEW

**Comment by bestony** [comment]
@SukkaW 

- [x]  Docker image to avoid the environment setting issue.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,5 @@
+FROM node:10
```
Comment (SukkaW) [inline Dockerfile]: Is it possible to use `node-alpine` to shrink the size of the image?

File: hexod [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
+docker run \
+  --interactive --tty --rm \
+  --volume "$PWD":/hexosite \
+  --workdir /hexosite \
+  -p 4000:4000 \
+  bestony/hexojs:latest "$@"
```
Comment (SukkaW) [inline hexod]: Is it possible to wrap `hexo` inside so that users can run `hexod [command] [--args]` directly?

File: Dockerfile (hunk 1 above)
Comment (bestony) [reply Dockerfile]: It's OK, But `node-alpine` maybe lost some  performance

File: hexod (hunk 2 above)
Comment (bestony) [reply hexod]: Of course yes, I will push a new commit  again.

**Comment by bestony** [comment]
@SukkaW 

Done with https://github.com/hexojs/hexo/pull/3891/commits/5fe7235d63b69af6d5999ccbd51a0cb8a60d51a7

Usage:

1.  `hexod init <name>`
2. `hexod version`
 
&& new Image only  112 MB (compare with 912 MB for `node:10`)

File: Dockerfile [hunk 3]
```diff
@@ -1,5 +1,7 @@
-FROM node:10
+FROM node:current-alpine
 
+RUN apk update && apk add bash git
```
Comment (SukkaW) [inline Dockerfile]: Use `RUN apk --no-cache add git` to keep the image as slim as possible.
Also, Alpine has built-in shell "Almquist Shell" (`/bin/ash`) which I believe is quite enough for execute hexo-cli.

Comment (bestony) [reply Dockerfile]: Tks! I will replace it.

**Comment by bestony** [comment]
@SukkaW Done.
New Image with ash  only 107MB.

**Comment by curbengh** [comment]
> is [hexo](https://hub.docker.com/u/hexo) belongs to hexo?

I don't know about this. cc @NoahDragon

**Comment by bestony** [comment]
if no one have the `hexo` organization, I create a `hexojs` organization, I can push to this organ namespace

**Comment by SukkaW** [comment]
@bestony I have realized something: wrap `hexo` in `hexod` looks simpler, but npm will become inaccessible. Maybe another wrapper `npmd` for npm as well?

**Comment by bestony** [comment]
It's ok, But it must have to file to link docker command.
maybe add an alias to hexod is better?
just like

```
Hexod init  -> hexo init
hexod npm install -> npm install 
```
it maybe better than npmd ?

**Comment by SukkaW** [comment]
@bestony Agree, LGTM. The original `hexod` could handle this well.

**Comment by bestony** [comment]
ok, I will try to add this feature

**Comment by curbengh** [comment]
> if no one have the `hexo` organization, I create a `hexojs` organization, I can push to this organ namespace
> 
> link: https://hub.docker.com/orgs/hexojs

I just created a docker account, feel free to add me https://hub.docker.com/u/curben

**Comment by SukkaW** [comment]
> I just created a docker account, feel free to add me https://hub.docker.com/u/curben

And this is mine: https://hub.docker.com/u/sukka

**Comment by SukkaW** [comment]
IMHO, we should close this PR and create a new repo then.

`Dockerfile` is no need to be published to npm. So it is unnecessary to have it in the repo. Also, Docker image is built with `npm install` instead of relying pm any files under `hexojs/hexo`.

@bestony What about creating a new GitHub repo `hexo-docker` and then transfer it to `hexojs` org then?

**Comment by bestony** [comment]
good idea. i will create a new repo,and try to transfer to hexojs

**Comment by bestony** [comment]
@SukkaW @curbengh Hi, I create a new repo for this pr, and i want to transfer it to hexojs. but alert "You don’t have the permission to create public repositories on hexojs".

so, what should i do next ?

My repo is here -> https://github.com/bestony/hexo-docker

**Comment by bestony** [comment]
and i add github action to automate publish docker images, and now ,you can download images by  `docker run hexojs/hexojs` 

https://hub.docker.com/repository/docker/hexojs/hexojs

**Comment by SukkaW** [comment]
@NoahDragon Would you please add @bestony to Hexo's plugin team? So he could transfer his repo to @hexojs's org.

**Comment by bestony** [comment]
any one can check this issue.

**Pull request closed without merging**